module/zfs/dmu_send.c

```diff
@@ -3720,6 +3746,38 @@ dmu_recv_stream(dmu_recv_cookie_t *drc, vnode_t *vp, offset_t *voffp,
 	}
 	mutex_exit(&rwa->mutex);
 
+	/*
+	 * if we are receiving a full stream as clone, we need to free leftover
+	 * objects after the last one referenced in the stream.
```

Comment:
I think you mean largest one referenced in the stream, not last.

Reply:
leftover objects with IDs greater than the maximum ID referenced in the stream.

?

Reply:
Yeah, how about something like this to make it perfectly clear.

"If we are receiving a full stream as a clone, all object ids which
are greater than the maximum ID referenced in the stream are
by definition unused and must be freed."